unit_tests/rms_tests.cc

```diff
@@ -0,0 +1,81 @@
+/* daft engineers vocoder project
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+#include "../include/rms/rms.hh"
+#include <cmath>
+#include <gtest/gtest.h>
+#include <mutex>
+#include <thread>
+
+// This is a unit test, magic numbers are called test data here
+// NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers)
+TEST(RMSTests, buffer) {
+    Pipe<Audio> input{};
+    Pipe<double> output{};
+    rms::RMS<4> rms(input, output);
+    Audio audio{0};
+
+    // check that rms is 0 at the beginning
+    ASSERT_EQ(rms.calc(), 0);
+
+    audio[0] = 16;
+    rms.insert(audio);
+    // buffer should now be 256, 0, 0, 0 giving rms of 8
+    ASSERT_EQ(rms.calc(), 8);
+
+    audio[0] = -16;
+    rms.insert(audio);
+    // buffer should now be 256, 256, 0, 0 giving rms of 8r2
+    ASSERT_EQ(rms.calc(), 8 * std::sqrt(2));
+
+    rms.insert(audio);
+    rms.insert(audio);
+    // buffer should now be 256, 256, 256, 256 giving rms of 16
+    ASSERT_EQ(rms.calc(), 16);
+
+    // insertions to the buffer from here should roll back to the front
+
+    audio[0] = 0;
+
+    for (int i = 0; i < 4; i++) {
+        rms.insert(audio);
+    }
+
+    // buffer should now be 0, 0, 0, 0, giving rms of 0
+    ASSERT_EQ(rms.calc(), 0);
+}
+
+// TEST macro violates guidelines
+// NOLINTNEXTLINE(cppcoreguidelines-owning-memory, cppcoreguidelines-avoid-non-const-global-variables)
+TEST(RMSTests, Integration) {
+    Pipe<Audio> input{};
+    Pipe<double> output{};
+    rms::RMS<4> rms(input, output);
+
+    rms.run();
+
+    std::thread input_thread{[&input] {
+        auto *pipe = &input;
```

Comment:
pointers are banned, can you change this to a reference

Reply:
fixed